OpenRA.Mods.Common/Traits/World/Locomotor.cs

```diff
@@ -288,31 +290,37 @@ bool CanMoveFreelyInto(Actor actor, Actor ignoreActor, CPos cell, CellConditions
 			if (cellBlocking == CellBlocking.Empty)
 				return true;
 
+			if (sharesCell && cellBlocking.HasCellBlocking(CellBlocking.FreeSubCell))
+				return true;
+
 			// We are blocked
-			if (ignoreActor == null && blocking.Overlaps(actor.Owner.PlayerMask))
+			if (ignoreActor == null && blocking.Overlaps(actor.Owner.PlayerMask) && cellBlocking < CellBlocking.Crushable)
 				return false;
 
-			if (cellBlocking.HasCellBlocking(CellBlocking.Crushable))
-				return true;
+			if (ignoreActor == null && check == CellConditions.BlockedByMovers && cellBlocking < CellBlocking.Crushable)
+				return false;
 
-			if (sharesCell && cellBlocking.HasCellBlocking(CellBlocking.FreeSubCell))
+			// We can only do an early exit here when we know the unit is not moving.
+			// Because we cant distinguish between crushable and ally moving (which we need to fall back to the old code)
```

Comment:
Maybe I'm still not understanding the logic here, but shouldn't this be something like
```csharp
// Moving actors have direction-dependent blocking, which must be evaluated on the slow path
// unless we know we can crush the actor (crushing is direction independent)
if (!blocking.Overlaps(actor.Owner.PlayerMask) && (!cellBlocking.HasCellBlocking(CellBlocking.HasMovingActor) || cellBlocking.HasCellBlocking(CellBlocking.Crushable))
	return true;
```

If I'm indeed not understanding the logic that makes it even more important to have a clear comment!